pkg/kubelet/util/csr/csr.go

```diff
@@ -50,67 +60,123 @@ func RequestNodeCertificate(client certificatesclient.CertificateSigningRequestI
+			return nil, fmt.Errorf("retrieved csr is not compatible: %v", err)
+		}
+	default:
+		return nil, fmt.Errorf("cannot create certificate signing request: %v", err)
 	}
-	defer certWatch.Stop()
-	ch := certWatch.ResultChan()
 
-	for {
-		event, ok := <-ch
-		if !ok {
-			break
-		}
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", req.Name).String()
 
-		if event.Type == watch.Modified || event.Type == watch.Added {
-			if event.Object.(*certificates.CertificateSigningRequest).UID != req.UID {
-				continue
+	event, err := cache.ListWatchUntil(
+		3600*time.Second,
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = fieldSelector
+				return client.List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = fieldSelector
+				return client.Watch(options)
+			},
+		},
+		func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Modified, watch.Added:
+			default:
+				return false, nil
+			}
+			csr := event.Object.(*certificates.CertificateSigningRequest)
+			if csr.UID != req.UID {
+				return false, fmt.Errorf("csr %q changed UIDs", csr.Name)
 			}
-			status := event.Object.(*certificates.CertificateSigningRequest).Status
-			for _, c := range status.Conditions {
+			for _, c := range csr.Status.Conditions {
 				if c.Type == certificates.CertificateDenied {
-					return nil, fmt.Errorf("certificate signing request is not approved, reason: %v, message: %v", c.Reason, c.Message)
+					return false, fmt.Errorf("certificate signing request is not approved, reason: %v, message: %v", c.Reason, c.Message)
 				}
-				if c.Type == certificates.CertificateApproved && status.Certificate != nil {
-					return status.Certificate, nil
+				if c.Type == certificates.CertificateApproved && csr.Status.Certificate != nil {
+					return true, nil
 				}
 			}
-		}
+			return false, nil
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("cannot watch on the certificate signing request: %v", err)
+	}
+
+	return event.Object.(*certificates.CertificateSigningRequest).Status.Certificate, nil
+
+}
+
+// This digest should include all the relevant pieces of the CSR we care about.
+// We can't direcly hash the serialized CSR because of random padding that we
+// regenerate every loop and we include usages which are not contained in the
+// CSR. This needs to be kept up to date as we add new fields to the node
+// certificates and with ensureCompatible.
+func digestedName(privateKeyData []byte, subject *pkix.Name, usages []certificates.KeyUsage) string {
+	hash := sha512.New512_256()
+	hash.Write(privateKeyData)
+	hash.Write([]byte(subject.CommonName))
+	for _, v := range subject.Organization {
+		hash.Write([]byte(v))
 	}
+	for _, v := range usages {
+		hash.Write([]byte(v))
+	}
+	return "node-csr-" + base64.URLEncoding.EncodeToString(hash.Sum(nil))
```

Comment:
Doesn't RawURLEncoding omit padding? Are all base64 characters allowed in CSR names?